compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -2825,7 +2831,7 @@ impl<'tcx> Ty<'tcx> {
 
             ty::Tuple(tys) => tys.iter().all(|ty| ty.is_trivially_sized(tcx)),
```

Comment:
Maybe there should be some threshold to not go too deep into nested types, or not branch too wide into huge tuples?

Reply:
If we are allowed to assume the type is WF, then we technically only need to check the last field of the tuple.

Reply:
How would I figure out if we can assume the Ty is WF?

Reply:
@lcnr has `is_trivially_sized` assume the type is WF or is it somehow itself involved in WF-checking? The question is whether we can determine that a duple is sized simply by checking its last field.

Reply:
good question, I think generally no, though I can't think of a way this can cause unsoundness with `Sized`. Opened a zulip thread to dump a few more thoughts here https://rust-lang.zulipchat.com/#narrow/stream/144729-t-types/topic/impls.20candidate.20assume.20wf

Reply:
I asked an "or" question and I don't know what your "no" answer means. :)

Reply:
which is why I've edited my comment :grin:

Reply:
I still don't know which side of the "or" you mean...

Reply:
what `is_trivially_sized` is doing for tuples is a simplified version of the builtin implementation of `Sized`. Ignoring constraints like "types cannot be named", builtin impls still have to follow all the rules which hold for user-written impls.

> can `is_trivially_sized` assume the type is WF

I believe that it must *generally not* assume the type is WF. This means it has to check all its fields to be `Sized`, not only the last one.

This is all a bit of a mess because correctly handling this requires the new approach to coinduction: the actual builtin impl currently only checks that the last field is `Sized`, meaning that changing `is_trivially_sized` to only check the last field would not have an impact on behavior. It would however make the switch to checking the tuple for WF when proving `Sized` to have a greater negative performance impact, which is :sweat_smile: :shrug:

Reply:
Okay, thanks for elaborating. Sounds like we want a `is_trivially_sized_assume_wf` then or so...